### PR TITLE
#388 Laos Create Company template

### PR DESCRIPTION
**Pull request by CarlosNZ**
First implementation here, still need to do:

- ~~pre-populate fields using defaults~~ (some done, others need further clarification across whole workflow)
- decide exactly which tables should be updated on Approval (currently ALL fields go to `organisation`, but some probably should go to `user`)
- permission_name "applyOrgLicence" doesn't actually exist yet, even though template grants it
- change `isLinear` to "true"
- have not created any org fields based on docs yet -- do we need to discuss how to handle?

Also done as part of this PR:

- add "01_users.js" to laos -- basically same as dev/users, but with generic applicant instead of our personal names
- updated the `/check-unique` endpoint so it can check any table/field combination (as well as existing "presets"). Also ensure it returns "true" if table or field doesn't exist yet
- bump evaluator package version to 1.8.0 in evaluator gui

### Testing:

- fresh install with `yarn database_init laos`
- log in as "applicant" (PW: 123456)
- Create company and submit
- Review as "userWithMultiplePermissions" -- Approve all and Submit
- Org should be created with all the new fields

### CODE REVIEW

**Comment by nmadruga** [comment]
> * permission_name "applyOrgLicence" doesn't actually exist yet, even though template grants it

Done in PR #396

**Comment by nmadruga** [comment]
> * have not created any org fields based on docs yet -- do we need to discuss how to handle?

Yes. Good point to discuss today after lunch

**Comment by nmadruga** [comment]
Just a warning: This PR has changes from 2 other PRs, so advisable to merge in this order:
- User registration changes #394 
- One of the other 2 PRs - which have each other merged - this or #396 (with updated changes from this one)

File: database/insertData/laos/02_template_orgRegistration.js [hunk 1]
```diff
@@ -0,0 +1,1003 @@
...
+                      isRequired: false
+                      parameters: {
+                        label: "Village"
+                        description: "(if applicable)"
+                        maxLength: 150
+                      }
+                    }
+                    {
+                      code: "addressDistrict"
+                      index: 75
```
Comment (nmadruga) [inline database/insertData/laos/02_template_orgRegistration.js:201]: Same index

File: src/server.ts [hunk 2]
```diff
@@ -78,41 +78,46 @@ const startServer = async () => {
...
-        table = 'organisation'
-        field = 'name'
+        tableName = 'organisation'
+        fieldName = 'name'
         break
       case 'orgRegistration':
-        table = 'organisation'
-        field = 'registration'
+        tableName = 'organisation'
+        fieldName = 'registration'
```
Comment (nmadruga) [inline src/server.ts:105]: Hmm I'm wondering what is this field now? WOuld this be also a generated one?

File: database/insertData/laos/02_template_orgRegistration.js [hunk 3]
```diff
@@ -0,0 +1,1003 @@
...
+              }
+              # Review Org rego
+              {
+                permissionNameToPermissionNameId: {
+                  connectByName: { name: "reviewOrgRego" }
+                }
+                stageNumber: 1
+                levelNumber: 1
+                canSelfAssign: true
+              }
```
Comment (nmadruga) [inline database/insertData/laos/02_template_orgRegistration.js:983]: Is the company that you can review been restricted already?

File: database/insertData/laos/02_template_orgRegistration.js [hunk 4]
```diff
@@ -0,0 +1,1003 @@
...
+              # Assign General
+              {
+                permissionNameToPermissionNameId: {
+                  connectByName: { name: "assignGeneral" }
+                }
+                stageNumber: 1
+                levelNumber: 1
+              }
```
Comment (nmadruga) [inline database/insertData/laos/02_template_orgRegistration.js:991]: Not sure why granting this one

File: database/insertData/laos/02_template_orgRegistration.js [hunk 5]
```diff
@@ -0,0 +1,1003 @@
...
+                parameterQueries: {
+                  username: {
+                    operator: "objectProperties"
+                    children: ["applicationData.username"]
+                  }
+                  orgName: {
+                    operator: "objectProperties"
+                    children: ["outputCumulative.organisation.name"]
+                  }
+                  permissionNames: ["reviewJoinOrg", "applyOrgLicense"]
```
Comment (nmadruga) [inline database/insertData/laos/02_template_orgRegistration.js:956]: Similar to my previous question for "reviewJoinOrg": 
Is the company that you can review been restricted already?

File: database/insertData/laos/02_template_orgRegistration.js [hunk 6]
```diff
@@ -0,0 +1,1003 @@
...
+                    }
+                    {
+                      code: "ownerName"
+                      index: 110
+                      title: "Company Owner"
+                      elementTypePluginCode: "shortText"
+                      category: QUESTION
+                      parameters: {
+                        label: "Full name of company owner"
+                        description: "TO-DO: Pre-populate"
```
Comment (nmadruga) [inline database/insertData/laos/02_template_orgRegistration.js:306]: I think should either remove this TODO - so this is just a regular text field for now - or enter the current user First name here and all next fields need to be captured from "User registration" which will need an update to also have Phone and fax

File: database/insertData/laos/02_template_orgRegistration.js [hunk 7]
```diff
@@ -0,0 +1,1003 @@
...
+                            children: ["responses.thisResponse"]
+                          }
+                          "^[0-9()\\\\-\+]*$"
+                        ]
+                      }
+                      validationMessage: "Not a valid phone number"
+                    }
+                    {
+                      code: "ownerEmail"
+                      index: 120
```
Comment (nmadruga) [inline database/insertData/laos/02_template_orgRegistration.js:334]: Same index

File: database/insertData/laos/02_template_orgRegistration.js [hunk 8]
```diff
@@ -0,0 +1,1003 @@
...
+                    {
+                      code: "S2Intro"
+                      index: 10
+                      title: "Intro Section 2"
+                      elementTypePluginCode: "textInfo"
+                      category: INFORMATION
+                      helpText: "The following questions are regarding the professional qualifications and experience of the **individual** who will primarily act on the company's behalf"
+                      parameters: {
+                        title: "## Professional details"
+                        text: "To-do: show Users name here with look-up"
```
Comment (nmadruga) [inline database/insertData/laos/02_template_orgRegistration.js:386]: Not sure if I understood this TODO

File: database/insertData/laos/02_template_orgRegistration.js [hunk 9]
```diff
@@ -0,0 +1,1003 @@
...
+                        inputFields: [
+                          {
+                            code: "PEtype"
+                            title: "Type of role"
+                            elementTypePluginCode: "radioChoice"
+                            category: QUESTION
+                            parameters: {
+                              label: "Type of role"
+                              options: ["Government", "Private Sector"]
+                              display: "inline"
```
Comment (nmadruga) [inline database/insertData/laos/02_template_orgRegistration.js:430]:
```suggestion
                              layout: "inline"
```

**Review by nmadruga — CHANGES_REQUESTED** [review]
Hello, haven't looked to the "Documents" section yet. But as you can see I tried going through the process of Changes Requested back to the Applicant and found a few small things - and a bigger one (#777) to be fixed in separated issues.

Please just have a look on these comments and I'll come back to this again today : )

File: database/insertData/laos/01_users.js [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+                {
+                  permissionNameToPermissionNameId: {
+                    connectByName: { name: "applyOrgRego" }
+                  }
+                }
+                {
+                  permissionNameToPermissionNameId: {
+                    connectByName: { name: "reviewJoinOrg" }
+                  }
+                }
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply database/insertData/laos/01_users.js]: Sure. 3897e308

File: database/insertData/laos/02_template_orgRegistration.js [hunk 11]
```diff
@@ -0,0 +1,1003 @@
...
+                              {
+                                operator: "objectProperties"
+                                children: ["applicationData.config.serverREST"]
+                              }
+                              "/check-unique"
+                            ]
+                          }
+                          ["table", "field", "value"]
+                          "organisation"
+                          "name_lao"
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply database/insertData/laos/02_template_orgRegistration.js]: Good catch. Now that it's just using regular "name" field, I can simplify it to one of the present types: d795a756

File: database/insertData/laos/02_template_orgRegistration.js (hunk 1 above)
Comment (CarlosNZ) [reply database/insertData/laos/02_template_orgRegistration.js:200]: Cool 4afa97af

File: src/server.ts (hunk 2 above)
Comment (CarlosNZ) [reply src/server.ts]: Yeah, good point, this isn't meaningful anymore, I'll remove it -- b56ed1e5

File: database/insertData/laos/02_template_orgRegistration.js (hunk 3 above)
Comment (CarlosNZ) [reply database/insertData/laos/02_template_orgRegistration.js:1017]: I don't think it has yet, we haven't added any more complex restrictions in policies for reviews yet.

File: database/insertData/laos/02_template_orgRegistration.js (hunk 4 above)
Comment (CarlosNZ) [reply database/insertData/laos/02_template_orgRegistration.js:1025]: Not granting it, just setting it as a required permission to make assignments on this template. (I know it's self-assignable, but thought it should be able to be assigned by another as well)

File: database/insertData/laos/02_template_orgRegistration.js (hunk 5 above)
Comment (CarlosNZ) [reply database/insertData/laos/02_template_orgRegistration.js:990]: https://github.com/openmsupply/application-manager-server/pull/400#discussion_r648794767

File: database/insertData/laos/02_template_orgRegistration.js (hunk 6 above)
Comment (CarlosNZ) [reply database/insertData/laos/02_template_orgRegistration.js]: I'm working on this now.

File: database/insertData/laos/02_template_orgRegistration.js (hunk 7 above)
Comment (CarlosNZ) [reply database/insertData/laos/02_template_orgRegistration.js:368]: b90572a9

File: database/insertData/laos/02_template_orgRegistration.js (hunk 8 above)
Comment (CarlosNZ) [reply database/insertData/laos/02_template_orgRegistration.js:420]: I just meant the the TextInfo field should show:
### Professional details
#### Applicant's Name

to indicate who we're filling out the info for.

I'll do this shortly

File: database/insertData/laos/02_template_orgRegistration.js (hunk 9 above)
Comment (CarlosNZ) [reply database/insertData/laos/02_template_orgRegistration.js]: Oops -- can't even follow my own parameter definitions 🙄

File: database/insertData/laos/02_template_orgRegistration.js (hunk 8 above)
Comment (CarlosNZ) [reply database/insertData/laos/02_template_orgRegistration.js]: Done now as part of 6b864f08

Looks like:
Screen Shot 2021-06-10 at 2 07 20 PM

**Comment by nmadruga** [comment]
Noticed while checking my PR, on Create Company section 1 Page 3:
Screen Shot 2021-06-10 at 3 30 01 PM

Since this is the company email, II don't think it should come pre filled and shouldn't compulsory.

**Comment by nmadruga** [comment]
Should this one (Page 4) be part of user rego?

Screen Shot 2021-06-10 at 3 30 01 PM

**Comment by CarlosNZ** [comment]
> Noticed while checking my PR, on Create Company section 1 Page 3:
> Screen Shot 2021-06-10 at 3 30 01 PM
> 
> Since this is the company email, II don't think it should come pre filled and shouldn't compulsory.

I think you mean Page 2 (Page 3 is the owner's email). But anyway, wouldn't most owners use their own email as the company contact? 

Have made it non-required for now.

**Comment by CarlosNZ** [comment]
> Should this one (Page 4) be part of user rego?
> Screen Shot 2021-06-10 at 3 32 30 PM

Yeah I guess, but shall we wait till we've merged both our templates and then change it on both at once?

**Comment by nmadruga** [comment]
> > Should this one (Page 4) be part of user rego?
> > Screen Shot 2021-06-10 at 3 32 30 PM
> 
> Yeah I guess, but shall we wait till we've merged both our templates and then change it on both at once?

I think you can change as part of this one.... Idk
I wished the company license wasn't merged with this, because to be this one is mergeable and user registration too. Just need to wait for finishing the Company license. If you can make this change and remove the merge with company license that would me nice. Because only license would need to be updated then and will make things easier

**Comment by CarlosNZ** [comment]
> I think you can change as part of this one.... Idk

Okay, I'll fill it with a default from "User" -- I'll assume the field name will be `phone` okay?

> I wished the company license wasn't merged with this, because to be this one is mergeable and user registration too. Just need to wait for finishing the Company license. If you can make this change and remove the merge with company license that would me nice. Because only license would need to be updated then and will make things easier

Why don't we just merge this one into yours and close this PR? (and you can merge UserReg straight to master now)

---

Okay, I've filled it with default from "user" and seems to be working (with manually adding a "phone" field). Is this what you meant ? 8e52acc0

**Comment by nmadruga** [comment]
Yep.

I think closing this will be ok

**Comment by CarlosNZ** [comment]
Closing, and we will merge this branch into #396

**Pull request closed without merging**